# Import Library yang dibutuhkan

Disini, kami menggunakan library sklearn.neighbors import KNeighborsClassifier untuk melakukan klasifikasi menggunakan K Nearest Neigbors

In [ ]:
import pandas as pd # Dataframe
import numpy as np
import math

# Model related

from sklearn.neighbors import KNeighborsClassifier # KNN
from sklearn.model_selection import train_test_split # Import train_test_split function
from sklearn import metrics #Import scikit-learn metrics module for accuracy calculation
from sklearn.svm import SVC # SVM


# Visualisasi
from sklearn import tree # Visualisasi Decision Tree
import matplotlib.pyplot as pl

# Upload Dataset

Dataset Rain in Australia berisi tentang pengamatan cuaca harian pada daerah - daerah di seluruh Australia yang berisi 145460.

Dataset ini memiliki variabel target yaitu RainTomorow yaitu prediksi hujan pada esok hari

In [ ]:
from google.colab import files
uploaded = files.upload() 

Saving weatherAUS.csv to weatherAUS.csv


In [ ]:
df = pd.read_csv("weatherAUS.csv")

In [ ]:
df.head()

Date Location  MinTemp  ...  Temp3pm  RainToday  RainTomorrow
0  2008-12-01   Albury     13.4  ...     21.8         No            No
1  2008-12-02   Albury      7.4  ...     24.3         No            No
2  2008-12-03   Albury     12.9  ...     23.2         No            No
3  2008-12-04   Albury      9.2  ...     26.5         No            No
4  2008-12-05   Albury     17.5  ...     29.7         No            No

[5 rows x 23 columns]

In [ ]:
for col in df.columns:
    df[col] = df[col].astype('category')
    df[col] = df[col].cat.codes

# Data Preprocessing

Mengubah data kategorikal menjadi data numerikal dengan menggunakan cat codes 

Inisiasi variable target yaitu RainTomorrow.

In [ ]:
df.dropna(inplace=True) # Drop missing values
df.describe(include='all')

X, y = df.drop(columns=['RainTomorrow']), df['RainTomorrow'] # Membagi data antara input (X) dan output (y)
X

Date  Location  MinTemp  MaxTemp  ...  Cloud3pm  Temp9am  Temp3pm  RainToday
0        396         2      213      262  ...        -1      224      262          0
1        397         2      153      284  ...        -1      227      287          0
2        398         2      208      290  ...         2      265      276          0
3        399         2      171      313  ...        -1      236      309          0
4        400         2      254      356  ...         8      233      341          0
...      ...       ...      ...      ...  ...       ...      ...      ...        ...
145455  3431        41      107      267  ...        -1      156      268          0
145456  3432        41      115      286  ...        -1      164      289          0
145457  3433        41      133      302  ...        -1      180      305          0
145458  3434        41      157      303  ...         2      206      304          0
145459  3435        41      228       -1  ...         8      205      253          0

[145460 rows x 22 columns]

# UJI COBA K-Nearest Neighbors

Pertama, inisialisasi variasi untuk knn_test. Diantaranya kami memvariasikan test size, n_neighbour, dan weights.

In [ ]:
knn_test = {
  'test_size': [0.1, 0.2, 0.3, 0.4, 0.5, 0.6],  
  'n_neighbor': [3, 5, 7, 9, 11, 13, 15, 17],
  'weights': ['uniform', 'distance'],
}

Inisialisasi K-Nearest Neigbors menggunakan looping dengan variasi yang telah diinisialisasi sebelumnya. Setelah itu diinisialisasi untuk print hasilnya.

In [ ]:
row_format ="{:>10}" * (len(knn_test) + 2)
print(row_format.format(*knn_test, 'akurasi', 'f1'))
acc_bfr = -math.inf
for i in knn_test['test_size']:
  for j in knn_test['n_neighbor']:
    for k in knn_test['weights']:
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=i, random_state=42)
        knn = KNeighborsClassifier(n_neighbors=j, weights=k)
        knn.fit(X_train, y_train)

        # print(y_test)
        pred = knn.predict(X_test)
        acc = metrics.accuracy_score(y_test, pred)
        f1 = metrics.f1_score(y_test, pred, average='micro')
        
        print(row_format.format(str(i), str(j), str(k), '%.3f' % acc, '%.3f' % f1))


 test_sizen_neighbor   weights   akurasi        f1
       0.1         3   uniform     0.812     0.812
       0.1         3  distance     0.817     0.817
       0.1         5   uniform     0.815     0.815
       0.1         5  distance     0.820     0.820
       0.1         7   uniform     0.816     0.816
       0.1         7  distance     0.821     0.821
       0.1         9   uniform     0.817     0.817
       0.1         9  distance     0.822     0.822
       0.1        11   uniform     0.813     0.813
       0.1        11  distance     0.819     0.819
       0.1        13   uniform     0.812     0.812
       0.1        13  distance     0.818     0.818
       0.1        15   uniform     0.814     0.814
       0.1        15  distance     0.819     0.819
       0.1        17   uniform     0.813     0.813
       0.1        17  distance     0.818     0.818
       0.2         3   uniform     0.810     0.810
       0.2         3  distance     0.815     0.815
       0.2         5   uniform 

# Variasi Parameter Terbaik

Menginisialisasi untuk mencari variasi parameter terbaik yang dihasilkan dari sebelumnya. Kemudian di print dengan parameter ( t_size': i, 'n_neighbors': j, 'weights': k, 'accuracy': acc, 'f1': f1)

In [ ]:
if(acc > acc_bfr):
          acc_bfr = acc
          visualization_knn = {'t_size': i, 'n_neighbors': j, 'weights': k, 'accuracy': acc, 'f1': f1}

print('best param: ', visualization_knn)

best param:  {'t_size': 0.6, 'n_neighbors': 17, 'weights': 'distance', 'accuracy': 0.8137059443604198, 'f1': 0.8137059443604198}


# KESIMPULAN

Variasi parameter yang didapatkan adalah data dengan data size 60%, menggunakan N Neighbors = 18, dengan weights = distance didapatkan akurasi = 0.8137059443604198